File: Heap-Sort/HS.scala
def compare[T](func:(T,T)=>Boolean)(a:T,b:T) = func(a,b)

val larger = compare((a:Int,b:Int) => a > b)_
val smaller = compare((a:Int,b:Int) => a < b)_

def swapTwo[T](xs:Array[T],a:Int,b:Int) = {
  val temp = xs(a)
  xs(a) = xs(b)
  xs(b) = temp
  xs
}

def heapify(func:(Int,Int)=>Boolean, nl:Array[Int], ind:Int) = {
  val parent = nl(ind)
  val leftind = ind * 2 + 1
  val left = if (leftind >= nl.length) Some(None) else Some(nl(leftind))
  val rightind = ind * 2 + 2
  val right = if (rightind >= nl.length) Some(None) else Some(nl(rightind))

  if (left.get == None)
    (nl, null)
  else if (func(parent,left.getOrElse(0).asInstanceOf[Int])
    && (if (right.get == None) true
    else func(parent,right.getOrElse(0).asInstanceOf[Int])))
    (nl, null)
  else if (func(left.getOrElse(0).asInstanceOf[Int],parent)
    && (if (right.get == None) true
    else func(left.getOrElse(0).asInstanceOf[Int],right.getOrElse(0).asInstanceOf[Int])))
    (swapTwo(nl,leftind,ind),leftind)
  else
    (swapTwo(nl,rightind,ind),rightind)

}

//println(heapify(larger, Array(4,1,3,2,16,9,10,14,8,7),3)._1.mkString("-"))

def parentInd(ind:Int) = if (ind < 0) -1 else ((ind - 1) / 2)

def buildHeap(func:(Int,Int)=>Boolean, arr:Array[Int], stack:Array[Int] = Array()):Array[Int] = {
  var stackInner = Array[Int]()
  if (stack.isEmpty) {
    stackInner = stackInner :+ parentInd(arr.length - 1)
  }else{
    stackInner = stack
  }

  if (stackInner(0) == -1)
    arr
  else {
    val temp = heapify(func,arr,stackInner.last)
    val tempStack:Array[Int] = if (temp._2 != null)
      stackInner :+ temp._2.asInstanceOf[Int]
    else Array(stackInner(0) - 1)
    buildHeap(func, temp._1, tempStack)
  }

}

print(buildHeap(larger, Array(4,1,3,2,16,9,10,14,8,7)).mkString("-"))

File: Heap-Sort/HS.clj
(defn heapify [func nl ind] ; return new nl and child index
  (let [parent (nth nl ind) ; nl cannot be empty
        leftind (inc (* 2 ind))
        left (if (>= leftind (count nl)) nil (nth nl leftind))
        rightind (+ 2 (* 2 ind))
        right (if (>= rightind (count nl)) nil (nth nl rightind))]
    (cond
      (nil? left) [nl nil]
      (and (func parent left)
           (if (nil? right) true (func parent right)))
      [nl nil]
      (and (func left parent)
           (if (nil? right) true (func left right)))
      [(assoc nl leftind parent ind left) leftind]
      :else
      [(assoc nl rightind parent ind right) rightind]
      )))

;(print (heapify >= [4 1 3 2 16 9 10 14 8 7] 4))

(defn parent-ind [ind]
  (if (= -1 ind) -1
      (-> ind dec (/ 2) int)))

(defn build-heap [func arr]
  (loop [arr arr
         stack [(parent-ind (dec (count arr)))]]
    ;(println arr stack)
    (cond
      (= -1 (first stack)) arr
      :else
      (let [[newarr newind] (heapify func arr (last stack))]
        (recur newarr
               (cond newind (conj stack newind)
                     :else
                     [(dec (first stack))]))))))

(println (build-heap >= [4 1 3 2 16 9 10 14 8 7]))

(defn heap-sort [func arr]
  (loop [result []
         arr arr]
    (cond (= [nil] arr) result
          :else
          (let [[this & remaining] (build-heap func arr)]
            (print remaining)
            (recur (conj result this)
                   (into [] (conj (butlast remaining) (last remaining))))))))

(println (heap-sort >= [4 1 3 2 16 9 10 14 8 7]))
(println (heap-sort <= [4 1 3 2 16 9 10 14 8 7]))
